### PR TITLE
Ruler enhancement POC - Part3

### PR DESCRIPTION
**Pull request by lchanmann**
Storing rule execution history: https://docs.google.com/document/d/1SyenzyRPf-iaBEIHO3b0K1IZ1gkaaWba7SrJjplEW_4/edit#heading=h.h2q17ycyjj7h

### CODE REVIEW

File: src/foam/nanos/ruler/Rule.js [hunk 1]
```diff
@@ -180,7 +192,9 @@
       javaCode: `
         getAsyncAction().applyAction(x, obj, oldObj, ruler);
         if ( ! getAfter()
-          && Operations.REMOVE != getOperation()
+          && Operations.CREATE == getOperation()
+          || Operations.UPDATE == getOperation()
+          || Operations.CREATE_OR_UPDATE == getOperation()
         ) {
           ruler.getDelegate().put_(x, obj);
```
Comment (jlhughes) [inline src/foam/nanos/ruler/Rule.js:199]: this is an asynchronous, applyAction should return a promise so you can perform the getDelegate.put on it's completion.  Or does applyAction already do this when it completes?

File: src/foam/nanos/ruler/RuleEngine.java [hunk 2]
```diff
@@ -61,18 +62,66 @@ public void stop() {
...
         for (Rule rule : rules) {
           if ( stops_.get() ) return;
 
-          if ( rule.f(getX(), obj, oldObj) ) {
+          currentRule_ = rule;
+          if ( rule.f(getX(), obj, oldObj)
+            && rule.getAsyncAction() != null
+          ) {
             rule.asyncApply(getX(), obj, oldObj, RuleEngine.this);
+            saveHistory(rule, obj);
```
Comment (jlhughes) [inline src/foam/nanos/ruler/RuleEngine.java:103]: saveHistory saves the result of the rule, but with async, it's not complete yet.  Believe apply should saveHistory on completion.

File: src/foam/nanos/ruler/Rule.js (hunk 1 above)
Comment (lchanmann) [reply src/foam/nanos/ruler/Rule.js:199]: This block will be run inside a thread pool on the server side. https://github.com/lchanmann/foam2/blob/RulerEnhancementPOC_part3/src/foam/nanos/ruler/RuleEngine.java#L102.

File: src/foam/nanos/ruler/RuleEngine.java (hunk 2 above)
Comment (lchanmann) [reply src/foam/nanos/ruler/RuleEngine.java:103]: This is inside the thread pool and it would complete as is.

**Review by jlhughes — APPROVED** [review]

**Pull request merged**